internal/cmd/analyze/analyze.go

```diff
@@ -0,0 +1,622 @@
+func runtimeComplexityRule(m *ast.Module, analyzeResults []*analyzeResult, missing map[string][]string) map[string][]string {
+	analyzeResultsCopy := make([]*analyzeResult, len(analyzeResults))
+	copy(analyzeResultsCopy, analyzeResults)
+
+	complexityResults := make(map[string][]string)
+	var ruleAnalyzeResult analyzeResult
+
+	// walk over the rules, to calculate the time complexity
+	ast.WalkRules(m, func(r *ast.Rule) bool {
+		var complexity string
+
+		// check if rule has body
+		if len(r.Body) == 0 {
+			complexity = "O(1)"
+		} else {
+			// TimeComplexity(rule) = TimeComplexity(body)
+			// TODO: Make this a map to optimize rule lookup
+			i := 0
+			var item *analyzeResult
+			for i, item = range analyzeResultsCopy {
+				if item.Rule == r.Head.Name.String() {
+					ruleAnalyzeResult = *item
+					break
+				}
+			}
+
+			analyzeResultsCopy = append(analyzeResultsCopy[:i], analyzeResultsCopy[i+1:]...)
+			complexity = runtimeComplexityBody(r.Body, &ruleAnalyzeResult, analyzeResults)
+		}
+
+		if len(ruleAnalyzeResult.Missing) == 0 {
+			complexityResults[r.Head.Name.String()] = append(complexityResults[r.Head.Name.String()], strings.TrimSpace(complexity))
+		} else {
+			missing[r.Head.Name.String()] = ruleAnalyzeResult.Missing
+		}
+
+		return false
+	})
+
+	return complexityResults
+}
+
+// runtimeComplexityBody analyzes the expressions in the rule body and returns
+// the runtime complexity for the rule
+func runtimeComplexityBody(body ast.Body, analyzeRuleResult *analyzeResult, all []*analyzeResult) string {
+
+	results := []analyzeExpression{}
+	for _, expr := range body {
+		result := analyzeExpression{}
+		result.Raw = expr
+		runtimeComplexityExpr(expr, &result, analyzeRuleResult, all)
+		results = append(results, result)
+	}
+
+	// stitch the time complexity results for the rule by walking over the
+	// expressions in the rule body
+	var resultComplexity string
+
+	for i := len(results) - 1; i >= 0; i-- {
+
+		if results[i].TimeComplexity == "" {
+			continue
+		}
+		if len(resultComplexity) == 0 && results[i].TimeComplexity != "O(1)" {
+			resultComplexity += fmt.Sprintf("%v", results[i].TimeComplexity)
+		} else {
+			if ast.ContainsRefs(results[i].Raw) {
+				if results[i].TimeComplexity != "O(1)" {
+					resultComplexity = fmt.Sprintf("[ %v * %v ]", results[i].TimeComplexity, resultComplexity)
+				}
+
+			} else {
+				if results[i].TimeComplexity != "O(1)" {
+					resultComplexity = fmt.Sprintf("[ %v + %v ]", results[i].TimeComplexity, resultComplexity)
+				}
+			}
+		}
+	}
+
+	if resultComplexity == "" {
+		resultComplexity = "O(1)"
+	}
+
+	// remove end brackets
+	resultComplexity = strings.TrimPrefix(resultComplexity, "[")
+	return strings.TrimSuffix(resultComplexity, "]")
+}
+
+// runtimeComplexityExpr analyzes an expression and returns its runtime complexity
+func runtimeComplexityExpr(expr *ast.Expr, analyzeExprResult *analyzeExpression, analyzeRuleResult *analyzeResult, all []*analyzeResult) {
```

Comment:
nit: Can some of these bigger functions be split up into some helper functions? Can maybe separate the code that does the condition checks for type of expression/term and the handling of the types.